File: 백준/Gold/1987. 알파벳/알파벳.py
import sys 
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)
input = lambda: sys.stdin.readline().rstrip('\r\n')

dx, dy = (1, -1, 0, 0), (0, 0, 1, -1)

def main():
    r, c = map(int, input().split())
    board = [list(map(lambda x: ord(x)-65, input())) for _ in range(r)]
    s = {(0, 0, 1<<board[0][0])}
    maxdist = 1
    while s:
        x, y, bit = s.pop()
        if maxdist == 26: break
        for dir in range(4):
            nx, ny = x+dx[dir], y+dy[dir]
            if 0 <= nx < r and 0 <= ny <c and not bit & 1<<board[nx][ny]:
                nbit = bit|1<<board[nx][ny]
                maxdist = max(maxdist, nbit.bit_count())
                s.add((nx, ny, nbit))
    print(maxdist)
    return
if __name__ == '__main__':
    main()

File: 백준/Gold/1987. 알파벳/알파벳.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;

const int dx[] = {1, -1, 0, 0}, dy[] = {0, 0, 1, -1};
int r, c, maxdist = 1;
vector<vector<int>> board;

void dfs(int x, int y, int bit, int dist){
    maxdist = max(maxdist, dist);
    if(maxdist == 26) return;
    for(int dir = 0; dir < 4; dir++){
        int nx = x+dx[dir], ny = y+dy[dir];
        if(0 <= nx && nx < r && 0 <= ny && ny < c && !(bit&1<<board[nx][ny])){
            int nbit = bit|1<<board[nx][ny];
            dfs(nx, ny, nbit, dist+1);
        }
    }
}

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    cin >> r >> c; board.resize(r, vector<int>(c));
    for(int i = 0; i < r; i++){
        string row; cin >> row;
        for(int j = 0; j < c; j++) board[i][j] = row[j]-'A';
    }
    dfs(0, 0, 1<<board[0][0], 1);
    cout << maxdist;
    return 0;
} 